src/ast/rewriter/bit_blaster/bit_blaster_adder.cpp

```diff
@@ -0,0 +1,112 @@
+#include "ast/rewriter/bit_blaster/bit_blaster_adder.h"
+
+bit_blaster_adder::bit_blaster_adder(bool_rewriter & rewriter, unsigned sz, numeral const & value):
+    m_rewriter(rewriter),
+    m_constant(value)
+{
+    reduce();
+    expr_ref_vector empty(m());
+    m_variable.resize(sz, empty);
+}
+
+bit_blaster_adder::bit_blaster_adder(bool_rewriter & rewriter, unsigned sz, expr * const * bits):
+    bit_blaster_adder(rewriter, sz)
+{
+    for (unsigned i = 0; i < sz; i++)
+        add_bit(i, bits[i]);
+}
+
+expr_ref bit_blaster_adder::sum_bits(vector< expr_ref_vector > & columns, expr_ref_vector & out_bits) const {
+    SASSERT(out_bits.empty());
+
+    expr_ref_vector carries(m());
+    expr_ref tmp1(m()), tmp2(m()), tmp3(m());
+
+    for (auto & column : columns) {
+        column.append(carries);
+        carries.reset();
+
+        while (column.size() >= 3) {
+            expr * a = column.back();
```

Comment:
in principle, the column owns the reference count on expressions. Since you pop and push expressions on the column the reference count gets stale at times.
That is, after popping expr * a, you add tmp2 to the back of column, in the next iteration you pop tmp2 and shrink the column. 
The reference to tmp2, or a, is then stale after the pop_back(). 
It might be valid still because tmp2 is declared outside of the loop. 
Though this seems too brittle and subtle when later having to revisit this code.

We could simply delay popping the column and update it in place by maintaining a pointer to the current virtual size.

So:

unsigned sz = columns.size();
while (sz >= 3) {
    expr* a = columns.get(sz -1);
    expr* b = columns.get(sz - 2);
    expr* c = columns.get(sz - 3);
    tmp1 = mk_xor(a, b);
    tmp2 = mk_xor(tmp1, c);
    carries.push_back(mk_or(mk_and(a, b), mk_and(tmp1, c)));
    columns[sz-3] = tmp2;
    sz -= 3;
}

etc.

Reply:
I've fixed this by changing a/b/c from `expr *` to `expr_ref`. That seems to have fixed many of the regression tests that failed before, so I guess the reference counting bug was a key source of the failures.